Added the Hyperlink-Induced Topic Search (HITS) algorithm in GraphX

rfurman opened the pull request.

mengxr commented:
@rfurman Thanks for implementing HITS! Some high-level comments:

1. Please check the Spark code style guide: 
  https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide
  Some code doesn't follow the guide closely, e.g., spacing and parentheses.

2. There are some unnecessary caching and materialization code. Could you run some experiments on a medium dataset and see which RDDs should be cached?

3. Should we support warm start?

Please also find some minor comments inline.

mengxr commented on graphx/src/main/scala/org/apache/spark/graphx/lib/HITS.scala:
```diff
@@ -0,0 +1,133 @@
+
+import org.apache.spark.Logging
+import org.apache.spark.graphx._
+import org.apache.spark.rdd._
```
Please make imports more explicit.

mengxr commented on graphx/src/main/scala/org/apache/spark/graphx/lib/HITS.scala:
```diff
@@ -0,0 +1,133 @@
+
+    // Initialize the HITS graph with each edge attribute having
+    // its original attribute and each vertex with attribute 1.0
+    var hubGraph = graph.mapVertices( (id, attr) => 1.0 ).cache
```
Both `cache` and `cache()` are used in this PR. Please follow the Spark code style.

mengxr commented on graphx/src/main/scala/org/apache/spark/graphx/lib/HITS.scala:
```diff
@@ -0,0 +1,133 @@
+      authGraph.unpersistVertices(false)
+      authGraph = graph.outerJoinVertices(authUpdates) {
+        (id, auth, msgSum) => msgSum.getOrElse(0.0)/authScale
+      }.cache()
```
Is caching needed?

mengxr commented on graphx/src/main/scala/org/apache/spark/graphx/lib/HITS.scala:
```diff
@@ -0,0 +1,133 @@
+      authGraph = graph.outerJoinVertices(authUpdates) {
+        (id, auth, msgSum) => msgSum.getOrElse(0.0)/authScale
+      }.cache()
+      authGraph.edges.foreachPartition(x => {}) // also materializes hitsGraph.vertices
```
minor: materialization could be done by `RDD.count()`. (Do we need to materialize it here?)

rfurman commented:
Thanks @mengxr, I have fixed up the style (except maybe for long method declarations) and added in warm start functionality with three options: ColdStart (default), WarmStartFromHubData, WarmStartFromAuthData
